pkg/pushsync/pushsync.go

```diff
@@ -153,141 +154,99 @@ func (ps *PushSync) handler(ctx context.Context, p p2p.Peer, stream p2p.Stream)
-			}
-			defer debit.Cleanup()
+		return debit.Apply()
+	}
 
-			// return back receipt
-			signature, err := ps.signer.Sign(chunkAddress.Bytes())
-			if err != nil {
-				return fmt.Errorf("receipt signature: %w", err)
-			}
+	if ch.Replicate {
 
-			receipt := pb.Receipt{Address: chunkAddress.Bytes(), Signature: signature, Nonce: ps.nonce}
-			err = w.WriteMsgWithContext(ctxd, &receipt)
+		ps.metrics.HandlerReplication.Inc()
+		defer func() {
 			if err != nil {
-				return fmt.Errorf("send receipt to peer %s: %w", p.Address.String(), err)
+				ps.metrics.HandlerReplicationErrors.Inc()
 			}
+		}()
 
-			err = debit.Apply()
-			return err
+		if !ps.radiusChecker.IsWithinStorageRadius(chunk.Address()) {
+			return ErrOutOfDepthStoring
 		}
-	}
 
-	// forwarding replication
-	storerNode := false
-	defer func() {
-		if !storerNode && ps.warmedUp() && ps.radiusChecker.IsWithinStorageRadius(chunkAddress) {
-			verifiedChunk, err := ps.validStamp(chunk, ch.Stamp)
-			if err != nil {
-				logger.Warning("forwarder, invalid stamp for chunk", "chunk_address", chunkAddress)
-				return
-			}
-			_, err = ps.storer.Put(ctx, storage.ModePutSync, verifiedChunk)
-			if err != nil {
-				logger.Warning("within depth peer's attempt to store chunk failed", "chunk_address", verifiedChunk.Address(), "error", err)
-			}
+		span, _, ctxd := ps.tracer.StartSpanFromContext(ctx, "pushsync-replication-storage", ps.logger, opentracing.Tag{Key: "address", Value: chunkAddress.String()})
+		defer span.Finish()
+
+		err = store(ctxd)
+		if err != nil {
+			err = fmt.Errorf("replication: %w", err)
 		}
-	}()
 
-	receipt, err := ps.pushToClosest(ctx, chunk, false, p.Address)
+		return err
+	}
+
+	receipt, err := ps.pushToClosest(ctx, chunk, false)
 	if err != nil {
 		if errors.Is(err, topology.ErrWantSelf) {
-			storerNode = true
 			ps.metrics.Storer.Inc()
-			chunk, err = ps.validStamp(chunk, ch.Stamp)
-			if err != nil {
-				return fmt.Errorf("pushsync storer invalid stamp: %w", err)
-			}
-
-			_, err = ps.storer.Put(ctx, storage.ModePutSync, chunk)
-			if err != nil {
-				return fmt.Errorf("chunk store: %w", err)
-			}
-
-			signature, err := ps.signer.Sign(ch.Address)
-			if err != nil {
-				return fmt.Errorf("receipt signature: %w", err)
-			}
-
-			// return back receipt
-			debit, err := ps.accounting.PrepareDebit(ctx, p.Address, price)
-			if err != nil {
-				return fmt.Errorf("prepare debit to peer %s before writeback: %w", p.Address.String(), err)
-			}
-			defer debit.Cleanup()
-
-			receipt := pb.Receipt{Address: chunkAddress.Bytes(), Signature: signature, Nonce: ps.nonce}
-			if err := w.WriteMsgWithContext(ctx, &receipt); err != nil {
-				return fmt.Errorf("send receipt to peer %s: %w", p.Address.String(), err)
-			}
-
-			return debit.Apply()
+			return store(ctx)
 		}
 
 		ps.metrics.Forwarder.Inc()
-
 		return fmt.Errorf("handler: push to closest: %w", err)
 	}
```

Comment:
is this `s.accounting.PrepareDebit` still needed here (few lines below)? it looks like `store` function is going to do debiting, if this is true then this will debit once again?

Reply:
that debit call is for forwarder nodes only